### PR TITLE
NAS-100068 / 12 / Move "Syslog" options from System: General to Advanced in API

### PR DESCRIPTION
**Pull request by sv0**

### CODE REVIEW

File: gui/api/resources.py [hunk 1]
```diff
@@ -3992,7 +3995,8 @@ def dehydrate(self, bundle):
                     '_vncweb_url': device_vncweb_url
                 })
             if bundle.obj.device_set.filter(dtype='VNC').exists():
-                vnc_port = bundle.obj.device_set.filter(dtype='VNC').values_list('attributes', flat=True)[0].get('vnc_port', 5900 + bundle.obj.id)
+                vnc_port = bundle.obj.device_set.filter(
+                    dtype='VNC').values_list('attributes', flat=True)[0].get('vnc_port', 5900 + bundle.obj.id)
```
Comment (william-gr) [inline gui/api/resources.py]: This will not be used in 12, please revert changes here to reduce the diff between master.

File: gui/system/forms.py [hunk 2]
```diff
@@ -1157,11 +1157,14 @@ def clean(self):
...
 
     def middleware_clean(self, data):
+        """TODO: add doc-string. What is it for?"""
 
         data.pop('reset_sed_password', None)
 
         if data.get('sed_user'):
             data['sed_user'] = data['sed_user'].upper()
+        if data.get('sysloglevel'):
+            data['sysloglevel'] = data['sysloglevel'].upper()
```
Comment (william-gr) [inline gui/system/forms.py]: Same here, lets please revert this changes.

File: gui/system/migrations/0052_move_syslog_from_settings_model_to_advanced_model.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('system', '0051_add_adv_syslogserver_field'),
+    ]
+
+    operations = [
+        migrations.RunPython(move_syslog, empty_reverse_migration),
```
Comment (william-gr) [inline gui/system/migrations/0052_move_syslog_from_settings_model_to_advanced_model.py]: Please use migrations.RunPython.noop

File: gui/system/migrations/0052_move_syslog_from_settings_model_to_advanced_model.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
+from django.db import migrations
```
Comment (william-gr) [inline gui/system/migrations/0052_move_syslog_from_settings_model_to_advanced_model.py]: Please squash all the migrations.

Also you're adding the field to Advanced but not removing from Settings.

File: gui/system/models.py [hunk 5]
```diff
@@ -329,6 +330,27 @@ class Advanced(Model):
         help_text=_("Global password to unlock SED disks."),
         verbose_name=_("SED Password"),
     )
+    adv_sysloglevel = models.CharField(
```
Comment (william-gr) [inline gui/system/models.py]: Should start with stg_

File: src/middlewared/middlewared/plugins/system.py [hunk 6]
```diff
@@ -121,6 +121,10 @@ class Config:
             Bool('anonstats'),
             Str('sed_user', enum=['USER', 'MASTER']),
             Str('sed_passwd', private=True),
+            Str('sysloglevel', enum=['F_EMERG', 'F_ALERT', 'F_CRIT', 'F_ERR',
+                                     'F_WARNING', 'F_NOTICE', 'F_INFO',
+                                     'F_DEBUG', 'F_IS_DEBUG']),
+            Str('syslogserver'),
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/system.py]: You will also need to change system.general to reroute the syslog settings to use system.advanced.

Also please add a deprecation warning over there when these settings are used in system.general.

**Comment by sv0** [comment]
> Also please add a deprecation warning over there when these settings are used in system.general.

I've added a deprecation warning.
But not sure if it should be a deprecation warning or just a message put to log with `self.logger.debug`

File: src/middlewared/middlewared/plugins/system.py [hunk 7]
```diff
@@ -962,6 +973,19 @@ def _get_index(country_columns, column):
 
         When `syslogserver` is defined, `sysloglevel` makes sure that logs matching the specified level are sent.
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/system.py]: Could you also please put a note in the docstring here about the deprecation?
Some along the lines these attributes are deprecated as of 11.3 and will be permanently moved for 12?

Comment (william-gr) [reply src/middlewared/middlewared/plugins/system.py]: It shows up in http://freenas_ip/api/docs/

File: gui/system/migrations/0050_move_syslog.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    operations = [
+        migrations.AddField(
+            model_name='advanced',
+            name='adv_sysloglevel',
+            field=models.CharField(choices=[('f_emerg', 'Emergency'), ('f_alert', 'Alert'), ('f_crit', 'Critical'), ('f_err', 'Error'), ('f_warning', 'Warning'), ('f_notice', 'Notice'), ('f_info', 'Info'), ('f_debug', 'Debug'), ('f_is_debug', 'Is_Debug')], default='f_info', help_text='Specifies which messages will be logged by server. INFO and VERBOSE log transactions that server performs on behalf of the client. f_is_debug specify higher levels of debugging output. The default is f_info.', max_length=120, verbose_name='Syslog level'),
+        ),
+        migrations.AddField(
+            model_name='advanced',
+            name='adv_syslogserver',
+            field=models.CharField(blank=True, default='', help_text='Specifies the server and port syslog messages will be sent to.  The accepted format is hostname:port or ip:port, if :port is not specified it will default to port 514 (this field currently only takes IPv4 addresses)', max_length=120, verbose_name='Syslog server'),
```
Comment (wblock) [inline gui/system/migrations/0050_move_syslog.py]: Use present tense ("is" or "are" rather than "will be").
```suggestion
            field=models.CharField(blank=True, default='', help_text='The server and port syslog messages are sent to.  The accepted format is hostname:port or ip:port. If :port is not specified, it defaults to port 514. This field currently only takes IPv4 addresses.', max_length=120, verbose_name='Syslog server'),
```

File: gui/system/models.py [hunk 9]
```diff
@@ -329,6 +308,27 @@ class Advanced(Model):
...
+        max_length=120,
+        choices=choices.SYS_LOG_LEVEL,
+        default="f_info",
+        verbose_name=_("Syslog level"),
+        help_text=_("Specifies which messages will be logged by "
```
Comment (wblock) [inline gui/system/models.py]:
```suggestion
        help_text=_("Which messages are logged by the "
```

File: gui/system/models.py [hunk 10]
```diff
@@ -329,6 +308,27 @@ class Advanced(Model):
...
+                    "server. INFO and VERBOSE log transactions that "
+                    "server performs on behalf of the client. "
```
Comment (wblock) [inline gui/system/models.py]:
```suggestion
                    "the server performs on behalf of the client. "
```

File: gui/system/models.py [hunk 11]
```diff
@@ -329,6 +308,27 @@ class Advanced(Model):
...
+                    "f_is_debug specify higher levels of debugging output. "
```
Comment (wblock) [inline gui/system/models.py]:
```suggestion
                    "f_is_debug specifies higher levels of debugging output. "
```

File: gui/system/models.py [hunk 12]
```diff
@@ -329,6 +308,27 @@ class Advanced(Model):
...
+                    "The default is f_info."),
+    )
+    adv_syslogserver = models.CharField(
+        default='',
+        blank=True,
+        max_length=120,
+        verbose_name=_("Syslog server"),
+        help_text=_("Specifies the server and port syslog messages "
```
Comment (wblock) [inline gui/system/models.py]: This is a duplicate of 0050_move_syslog.py and should be reworded the same way.

File: src/middlewared/middlewared/plugins/system.py [hunk 13]
```diff
@@ -98,6 +101,21 @@ class Config:
                         'Serial port specified has not been identified by the system'
```
Comment (wblock) [inline src/middlewared/middlewared/plugins/system.py]: Not clear what "identified" means.  Probably "is not present on this system."

File: src/middlewared/middlewared/plugins/system.py [hunk 14]
```diff
@@ -98,6 +101,21 @@ class Config:
...
+                verrors.add(
+                    f'{schema}.syslogserver',
+                    'Invalid syslog server format'
+                )
+            elif ':' in syslog_server:
+                port = int(syslog_server.split(':')[-1])
+                if port < 0 or port > 65535:
+                    verrors.add(
+                        f'{schema}.syslogserver',
+                        'Port specified should be between 0 - 65535'
```
Comment (wblock) [inline src/middlewared/middlewared/plugins/system.py]: "Should" is a recommendation, use "must" when it is required.
```suggestion
                        'Port must be in the range of 0 to 65535.'
```

File: src/middlewared/middlewared/plugins/system.py [hunk 15]
```diff
@@ -133,6 +155,8 @@ class Config:
 
         `autotune` when enabled executes autotune script which attempts to optimize the system based on the installed
         hardware.
+
+        When `syslogserver` is defined, `sysloglevel` makes sure that logs matching the specified level are sent.
```
Comment (wblock) [inline src/middlewared/middlewared/plugins/system.py]: Not clear what "makes sure" means here.  Sounds like sometimes it doesn't send them.

File: src/middlewared/middlewared/plugins/system.py [hunk 16]
```diff
@@ -960,8 +973,22 @@ def _get_index(country_columns, column):
         `ui_address` and `ui_v6address` are a list of valid ipv4/ipv6 addresses respectively which the system will
         listen on.
 
-        When `syslogserver` is defined, `sysloglevel` makes sure that logs matching the specified level are sent.
+        `syslogserver` and `sysloglevel` are deprecated fields as of 11.3 and will be permanently moved for 12.0
```
Comment (wblock) [inline src/middlewared/middlewared/plugins/system.py]: Moved to what?

File: src/middlewared/middlewared/plugins/system.py [hunk 17]
```diff
@@ -960,8 +973,22 @@ def _get_index(country_columns, column):
...
+
         """
+        advadced_config = {}
```
Comment (wblock) [inline src/middlewared/middlewared/plugins/system.py]: Typo.

File: src/middlewared/middlewared/plugins/system.py [hunk 18]
```diff
@@ -960,8 +973,22 @@ def _get_index(country_columns, column):
...
+        # fields were moved to Advanced
+        for deprecated_field in ('sysloglevel', 'syslogserver'):
+            if deprecated_field in data:
+                warnings.warn(
+                    f"{deprecated_field} has been deprecated and moved to 'system.advanced'",
+                    DeprecationWarning
+                )
+                advadced_config[deprecated_field] = data[deprecated_field]
```
Comment (wblock) [inline src/middlewared/middlewared/plugins/system.py]: Typo again.  Maybe this is intentional.

File: src/middlewared/middlewared/plugins/system.py [hunk 19]
```diff
@@ -960,8 +973,22 @@ def _get_index(country_columns, column):
...
+                del data[deprecated_field]
+        if advadced_config:
```
Comment (wblock) [inline src/middlewared/middlewared/plugins/system.py]: And again.

File: src/middlewared/middlewared/plugins/system.py [hunk 20]
```diff
@@ -960,8 +973,22 @@ def _get_index(country_columns, column):
...
+            await self.middleware.call('system.advanced.update', advadced_config)
```
Comment (wblock) [inline src/middlewared/middlewared/plugins/system.py]: Here too.

File: src/middlewared/middlewared/plugins/system.py (hunk 13 above)
Comment (william-gr) [reply src/middlewared/middlewared/plugins/system.py]: Not part of the PR so we can handle it in another PR/ticket.

File: src/middlewared/middlewared/plugins/system.py (hunk 15 above)
Comment (sv0) [reply src/middlewared/middlewared/plugins/system.py]: > Not clear what "makes sure" means here. Sounds like sometimes it doesn't send them.

What about:
"When `syslogserver` is defined, the logs matching the `sysloglevel` level are sent." ?

File: gui/system/migrations/0050_move_syslog.py [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+    operations = [
+        migrations.AddField(
+            model_name='advanced',
+            name='adv_sysloglevel',
+            field=models.CharField(choices=[('f_emerg', 'Emergency'), ('f_alert', 'Alert'), ('f_crit', 'Critical'), ('f_err', 'Error'), ('f_warning', 'Warning'), ('f_notice', 'Notice'), ('f_info', 'Info'), ('f_debug', 'Debug'), ('f_is_debug', 'Is_Debug')], default='f_info', max_length=120, verbose_name='Syslog level'),
+        ),
+        migrations.AddField(
+            model_name='advanced',
+            name='adv_syslogserver',
+            field=models.CharField(blank=True, default='', max_length=120, verbose_name='Syslog server'),
```
Comment (william-gr) [inline gui/system/migrations/0050_move_syslog.py]: Curious if removing help_text here wont make django think there are pending changes?

File: src/middlewared/middlewared/plugins/system.py [hunk 22]
```diff
@@ -960,8 +973,22 @@ def _get_index(country_columns, column):
         `ui_address` and `ui_v6address` are a list of valid ipv4/ipv6 addresses respectively which the system will
         listen on.
 
-        When `syslogserver` is defined, `sysloglevel` makes sure that logs matching the specified level are sent.
+        `syslogserver` and `sysloglevel` are deprecated fields as of 11.3
+        and will be permanently moved to SystemAdvancedService for 12.0
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/system.py]: Please replace SystemAdvancedService with system.advanced.update.
API readers do not know what SystemAdvancedService is.

File: gui/system/migrations/0050_move_syslog.py (hunk 21 above)
Comment (sv0) [reply gui/system/migrations/0050_move_syslog.py]: Did not experience such problem.
That `help_text` in the migration does not make a lot of sense in this particular case.

File: src/middlewared/middlewared/plugins/system.py (hunk 15 above)
Comment (william-gr) [reply src/middlewared/middlewared/plugins/system.py]: `syslog level or above`?

Comment (william-gr) [reply src/middlewared/middlewared/plugins/system.py]: @wblock thoughts?

File: gui/system/models.py (hunk 12 above)
Comment (sv0) [reply gui/system/models.py]: I've removed `help_text` from the migration.

File: src/middlewared/middlewared/plugins/system.py (hunk 15 above)
Comment (wblock) [reply src/middlewared/middlewared/plugins/system.py]: That is clearer.  
"When `syslogserver` is defined, logs of `sysloglevel` or above are sent."
